include/nbl/ext/TextRendering/TextRendering.h

```diff
@@ -0,0 +1,103 @@
+// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#ifndef _NBL_EXT_TEXT_RENDERING_H_INCLUDED_
+#define _NBL_EXT_TEXT_RENDERING_H_INCLUDED_
+
+#include "nabla.h"
+
+#include "nbl/video/alloc/SubAllocatedDataBuffer.h"
+#include "nbl/video/utilities/CPropertyPool.h"
+
+using namespace nbl;
+using namespace nbl::core;
+using namespace nbl::asset;
+using namespace nbl::video;
+
+namespace nbl
+{
+namespace ext
+{
+namespace TextRendering
+{
+
+struct SPixelCoord
+{
+	uint16_t x, y;
+};
+
+struct SGlyphData 
+{
+	SPixelCoord offsetFromStringMin;
+	uint32_t textureAtlasGlyphIndex;
+};
+
+struct StringBoundingBox
+{
+	SPixelCoord min, max;
+};
+
+struct SStringCreationParams
+{
+	SPixelCoord offset;
+	uint32_t glyphCount;
+	SGlyphData const* glyphs;
+};
+
+class NBL_API TextRenderer
+{
+public:
+	typedef typename uint32_t size_type;
+
+	TextRenderer(core::smart_refctd_ptr<ILogicalDevice> device, uint32_t maxGlyphCount, uint32_t maxStringCount);
+
+	template<typename... Args>
+	size_type allocateStrings(
+		uint32_t count,
+		size_type* outStrings,
+		const Args&&... args
+	);
+
+	void freeStrings(
+		uint32_t count,
+		const size_type* strings,
+		const size_type* stringGlyphCounts
+	);
+
+	// Creates the pipeline layout and the pipeline for rendering text
+	core::smart_refctd_ptr<video::IGPUGraphicsPipeline> createPipeline(
+		video::IGPUObjectFromAssetConverter::SParams& cpu2gpuParams,
+
+		// The data in these buffers is provided by the frustum culling system before calling drawText
+		core::smart_refctd_ptr<video::IGPUBuffer> visibleStrings,
+		core::smart_refctd_ptr<video::IGPUBuffer> visibleStringMvps,
+		core::smart_refctd_ptr<video::IGPUBuffer> visibleStringGlyphCount
+	);
```

Comment:
look at all the other extensions and utilities, they create default layouts in the constructor whenever feasible (pipeline and DS) and return cached versions

there are always separate functions for creating default layout descriptor sets and DS writing functions